store/streaming/plugins/abci/v1/grpc.go

```diff
@@ -0,0 +1,148 @@
+package v1
+
+import (
+	"context"
+	"os"
+
+	"github.com/hashicorp/go-plugin"
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var _ storetypes.ABCIListener = (*GRPCClient)(nil)
+
+// GRPCClient is an implementation of the ABCIListener interface that talks over RPC.
+type GRPCClient struct {
+	client ABCIListenerServiceClient
+}
+
+func (m GRPCClient) ListenBeginBlock(goCtx context.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	sm := ctx.StreamingManager()
+	request := &ListenBeginBlockRequest{Req: &req, Res: &res}
+
+	if sm.Async {
+		go func(ctx sdk.Context) {
+			_, err := m.client.ListenBeginBlock(ctx, request)
+			ctx.Logger().Error("BeginBlock listening hook failed", "height", ctx.BlockHeight(), "err", err)
```

Comment:
This should not execute if err == nil, but at the moment it does.

Reply:
Ah. I missed the check. 

I know we decided to add goroutines here for fire-and-forget but would it not be better to leave this up to the plugin implementation? This would also satisfy CI non-determinism checks (which it flagged). We can add a ctx `timeout` to mitigate any blocking concerns?

Reply:
If you want to leave this up to the implementation then you would need to remove the Async flag from the sm struct and always call ListenBeginBlock directly. In either case you should only call logger.Error if the err returned by the client call is non-nil.

Reply:
`async` functionality has been removed.